### PR TITLE
.Net: fixed extension data in Model diagnostics

### PR DESCRIPTION
**Pull request by TaoChenOSU**
### Motivation and Context

Previously when an AI client starts a model diagnostics activity, it passes in an execution setting that is not parsed to a setting that is specific to the client. This creates an issue where the some of the settings cannot be read by the diagnostics module.

### Description

Pass in the parsed setting to the diagnostics module. The diagnostics module will the serialize the object and deserialize it to `PromptExecutionSettings` to get the extension data.

image

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs [hunk 1]
```diff
@@ -294,8 +294,14 @@ private sealed class AIServiceSelector : IAIServiceSelector
                             Temperature = 0,
                             ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions
                         },
-                        GoogleAIGeminiServiceKey => new GeminiPromptExecutionSettings(),
-                        HuggingFaceServiceKey => new HuggingFacePromptExecutionSettings(),
+                        GoogleAIGeminiServiceKey => new GeminiPromptExecutionSettings()
```
Comment (alliscode) [inline dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs:297]: Is this change needed?

Comment (TaoChenOSU) [reply dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs:297]: If this is not set, it won't have a default and will be null. This makes sure the activity of the Gemini run has at least one execution setting in this sample app.

**Review by alliscode — APPROVED** [review]

File: dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs [hunk 2]
```diff
@@ -294,8 +294,14 @@ private sealed class AIServiceSelector : IAIServiceSelector
...
+                        {
+                            Temperature = 0,
+                        },
```
Comment (stephentoub) [inline dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs:300]: What's the reason for not including tool calling here?

File: dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceClient.cs [hunk 3]
```diff
@@ -134,7 +134,8 @@ internal HttpRequestMessage CreatePost(object requestData, Uri endpoint, string?
         var endpoint = this.GetTextGenerationEndpoint(modelId);
         var request = this.CreateTextRequest(prompt, executionSettings);
 
-        using var activity = ModelDiagnostics.StartCompletionActivity(endpoint, modelId, this.ModelProvider, prompt, executionSettings);
+        var huggingFaceExecutionSettings = HuggingFacePromptExecutionSettings.FromExecutionSettings(executionSettings);
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceClient.cs:137]: Should the LogTextGenerationUsage later in the method use this one instead of the original?

File: dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceMessageApiClient.cs [hunk 4]
```diff
@@ -85,7 +85,8 @@ internal sealed class HuggingFaceMessageApiClient
         var request = this.CreateChatRequest(chatHistory, executionSettings);
         request.Stream = true;
 
-        using var activity = ModelDiagnostics.StartCompletionActivity(endpoint, modelId, this._clientCore.ModelProvider, chatHistory, executionSettings);
+        var huggingFaceExecutionSettings = HuggingFacePromptExecutionSettings.FromExecutionSettings(executionSettings);
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceMessageApiClient.cs:88]: The original execution settings don't actually seem to be used anywhere in the implementation. Is that a bug? Or if it's by design, does it make sense to log information that's not actually used?

File: dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs [hunk 5]
```diff
@@ -109,16 +119,23 @@ public static bool IsModelDiagnosticsEnabled()
...
+    private static void AddOptionalTags<T>(Activity? activity, T? executionSettings) where T : PromptExecutionSettings
     {
-        if (activity is null || executionSettings?.ExtensionData is null)
+        if (activity is null || executionSettings is null)
+        {
+            return;
+        }
+
+        // Serialize and deserialize the execution settings to get the extension data
+        var deserializedSettings = JsonSerializer.Deserialize<PromptExecutionSettings>(JsonSerializer.Serialize(executionSettings));
```
Comment (stephentoub) [inline dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs:130]: Is this always going to do the right thing? i.e. the exact same names are used for all execution settings for MaxToken, Temperature, and TopP?

File: dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs [hunk 6]
```diff
@@ -194,13 +211,13 @@ private static void ToOpenAIFormat(StringBuilder sb, ChatMessageContentItemColle
     /// Start a completion activity and return the activity.
     /// The `formatPrompt` delegate won't be invoked if events are disabled.
     /// </summary>
-    private static Activity? StartCompletionActivity<T>(
+    private static Activity? StartCompletionActivity<T1, T2>(
```
Comment (stephentoub) [inline dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs:214]: How about TPrompt and TPromptExecutionSettings?

File: dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs (hunk 2 above)
Comment (TaoChenOSU) [reply dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs:300]: I found an issue with Gemini when invoking a Kernel function with auto invoke enabled. I will file an issue in a moment.

File: dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceMessageApiClient.cs (hunk 4 above)
Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceMessageApiClient.cs:88]: It's a bit strange in the HuggingFace connector. The original execution settings is being parsed inside the `CreateChatRequest` method and all the parameters are stored in the `request`.

I can actually parse the settings to `HuggingFacePromptExecutionSettings` outside of `CreateChatRequest` and pass in the result, to avoid repeating the operation.

File: dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs (hunk 2 above)
Comment (stephentoub) [reply dotnet/samples/Demos/TelemetryWithAppInsights/Program.cs:300]: Ok. I suggest we leave the line in but commented out and with the issue link next to it... it can then be uncommented when the issue is fixed (hopefully soon :smile:)

File: dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs (hunk 5 above)
Comment (TaoChenOSU) [reply dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs:130]: Yes, so far. We just need to be very careful about this. I am planning to add unit tests for all the model diagnostics code.

File: dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceClient.cs (hunk 3 above)
Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceClient.cs:137]: Good catch! Yes, in theory. But it won't change anything ATM. I will change it for consistency.

**Review by stephentoub — APPROVED** [review]
one unaddressed comment, otherwise lgtm

**Comment by TaoChenOSU** [comment]
> one unaddressed comment, otherwise lgtm

Added the line back with a comment and the link to the issue.

**Review by markwallace-microsoft — APPROVED** [review]
